src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenForEachTests.cs

```diff
@@ -2365,5 +2365,774 @@ .locals init (T V_0)
+    {
+        foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public struct Enumerator
+    {
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions
+{
+    public static C.Enumerator GetEnumerator(this in C self) => new C.Enumerator();
+}";
+            CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput: "123");
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaInExtensionOnAssignableVariable()
+        {
+            var source = @"
+using System;
+public struct C
+{
+    public static void Main()
+    {
+        var c = new C();
+        foreach (var i in c)
+        {
+            Console.Write(i);
+        }
+    }
+    public struct Enumerator
+    {
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions
+{
+    public static C.Enumerator GetEnumerator(this in C self) => new C.Enumerator();
+}";
+            CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput: "123");
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaExtensionsCSharp8()
+        {
+            var source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions
+{
+    public static C.Enumerator GetEnumerator(this C self) => new C.Enumerator();
+}";
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (7,27): error CS8652: The feature 'extension GetEnumerator' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         foreach (var i in new C())
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "new C()").WithArguments("extension GetEnumerator").WithLocation(7, 27)
+                );
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaInternalExtensions()
+        {
+            var source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions
+{
+    internal static C.Enumerator GetEnumerator(this C self) => new C.Enumerator();
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview)
```

Comment:
No idea why this doesn't work. `GetEnumerator` is accessible in `Main`, so this should be valid code.

Reply:
This is currently an error:

```csharp
using System.Collections.Generic;

public class C {
    public void M() {
        foreach(var a in this)
        {
        }
    }
    
    internal IEnumerator<int> GetEnumerator() => throw null;
}
```

GetEnumerator has to be public

I'll mention this on the open issue.

Reply:
Thanks, we'll see what LDM wants to do.

Reply:
It's already in the open issues issue, I don't think we need to duplicate it all around.